tokio/src/sync/notify.rs

```diff
@@ -0,0 +1,519 @@
+///             // Drain values
+///             if let Some(value) = self.values.lock().unwrap().pop_front() {
+///                 return value;
+///             }
+///
+///             // Wait for values to be available
+///             self.notify.notified().await;
+///         }
+///     }
+/// }
+/// ```
+///
+/// [park]: std::thread::park
+/// [unpark]: std::thread::Thread::unpark
+/// [`notified().await`]: Notify::notified()
+/// [`notify`]: Notify::notify()
+/// [`Semaphore`]: crate::sync::Semaphore
+#[derive(Debug)]
+pub struct Notify {
+    state: AtomicU8,
+    waiters: Mutex<LinkedList<Waiter>>,
+}
+
+#[derive(Debug)]
+struct Waiter {
+    /// Waiting task's waker
+    waker: Option<Waker>,
+
+    /// `true` if the notification has been assigned to this waiter.
+    notified: bool,
+}
+
+/// Future returned from `notified()`
+#[derive(Debug)]
+struct Notified<'a> {
+    /// The `Notify` being received on.
+    notify: &'a Notify,
+
+    /// The current state of the receiving process.
+    state: State,
+
+    /// Entry in the waiter `LinkedList`.
+    waiter: linked_list::Entry<Waiter>,
+}
+
+#[derive(Debug)]
+enum State {
+    Init,
+    Waiting,
+    Done,
+}
+
+/// Initial "idle" state
+const EMPTY: u8 = 0;
+
+/// One or more threads are currently waiting to be notified.
+const WAITING: u8 = 1;
+
+/// Pending notification
+const NOTIFIED: u8 = 2;
+
+impl Notify {
+    /// Create a new `Notify`, initialized without a permit.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::Notify;
+    ///
+    /// let notify = Notify::new();
+    /// ```
+    pub fn new() -> Notify {
+        Notify {
+            state: AtomicU8::new(0),
+            waiters: Mutex::new(LinkedList::new()),
+        }
+    }
+
+    /// Wait for a notification.
+    ///
+    /// Each `Notify` value holds a single permit. If a permit is available from
+    /// an earlier call to [`notify()`], then `notified().await` will complete
+    /// immediately, consuming that permit. Otherwise, `notified().await` waits
+    /// for a permit to be made available by the next call to `notify()`
+    ///
+    /// [`notify()`]: Notify::notify
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::Notify;
+    /// use std::sync::Arc;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let notify = Arc::new(Notify::new());
+    ///     let notify2 = notify.clone();
+    ///
+    ///     tokio::spawn(async move {
+    ///         notify2.notified().await;
+    ///         println!("received notification");
+    ///     });
+    ///
+    ///     println!("sending notification");
+    ///     notify.notify();
+    /// }
+    /// ```
+    pub async fn notified(&self) {
+        Notified {
+            notify: self,
+            state: State::Init,
+            waiter: linked_list::Entry::new(Waiter {
+                waker: None,
+                notified: false,
+            }),
+        }
+        .await
```

Comment:
I'm assuming we do this rather than just returning the `Notified` future because we don't want to commit to making it public API?

Reply:
yes... and docs.